hpx/util/stencil3_iterator.hpp

```diff
@@ -0,0 +1,227 @@
+                IterBegin const& begin, IterValueBegin const& begin_val,
+                IterEnd const& end, IterValueEnd const& end_val)
+            {
+                auto prev = make_previous_transformer(begin, begin_val);
+                auto next = make_next_transformer(end, end_val);
+
+                return type(util::make_tuple(
+                    make_transform_iterator(it, prev), it,
+                    make_transform_iterator(it, next)));
+            }
+
+            static type create(Iterator const& it)
+            {
+                return type(util::make_tuple(
+                    make_transform_iterator(it, left_transformer()), it,
+                    make_transform_iterator(it, right_transformer())));
+            }
+        };
+    }
+
+    template <typename Iterator, typename IterBegin, typename IterValueBegin,
+        typename IterEnd, typename IterValueEnd>
+    class stencil3_iterator
+      : public detail::stencil3_iterator_base<
+            Iterator, IterBegin, IterValueBegin, IterEnd, IterValueEnd
+        >::type
+    {
+    private:
+        typedef detail::stencil3_iterator_base<
+                Iterator, IterBegin, IterValueBegin, IterEnd, IterValueEnd
+            > base_maker;
+        typedef typename base_maker::type base_type;
+
+    public:
+        stencil3_iterator() {}
+
+        stencil3_iterator(Iterator const& it,
+                IterBegin const& begin, IterValueBegin const& begin_val,
+                IterEnd const& end, IterValueEnd const& end_val)
+          : base_type(base_maker::create(it, begin, begin_val, end, end_val))
+        {}
+
+        explicit stencil3_iterator(Iterator const& it)
+          : base_type(base_maker::create(it))
+        {}
+
+        template <typename ...Ts>
+        stencil3_iterator(tuple<Ts...> const& t)
+          : base_type(t)
+        {}
+
+    private:
+        friend class boost::iterator_core_access;
+
+        bool equal(stencil3_iterator const& other) const
+        {
+            return util::get<1>(this->get_iterator_tuple()) ==
+                   util::get<1>(other.get_iterator_tuple());
+        }
+    };
+
+    ///////////////////////////////////////////////////////////////////////////
+    template <typename Iterator, typename IterBegin, typename IterValueBegin,
+        typename IterEnd, typename IterValueEnd>
+    inline stencil3_iterator<
+        Iterator, IterBegin, IterValueBegin, IterEnd, IterValueEnd
+    >
+    make_stencil3_iterator(Iterator const& it,
+        IterBegin const& begin, IterValueBegin const& begin_val,
+        IterEnd const& end, IterValueEnd const& end_val)
+    {
+        typedef stencil3_iterator<
+                Iterator, IterBegin, IterValueBegin, IterEnd, IterValueEnd
+            > result_type;
+        return result_type(it, begin, begin_val, end, end_val);
+    }
+
+    template <typename StencilIterator, typename Iterator>
+    inline StencilIterator
+    make_stencil3_iterator(Iterator const& it)
```

Comment:
The purpose of this function is to create a new `stencil3_iterator` from the given iterator (`it`) for the underlying range `[begin, end)` and a pair of iterators referring to the values to be used as the boundary elements (`begin_val` and `end_val`).